### ISSUE TITLE
installing packages slow outside C:/ on Windows

### ISSUE THREAD

**Issue by mikmart** [rstudio/renv#201]
I have secondary drives that I usually do my project work on. If I create a project under C:/ everything works fine -- however, when I try to use renv with a project on any other drive, I run into a bunch of strange behaviour:

Package installation hangs for long periods of time after everything is seemingly done: the more packagese there are, the worse it gets. After large installations, like e.g. `install.packages("tidyverse")`, the R session hangs for several minutes.

Once installation eventually finishes, the packages are available, but don't get picked up by `renv::snapshot()` for updating the lock file. For example:

``` r
library(glue)
#> Error in library(glue) : there is no package called ‘glue’
install.packages("glue")
#> * Querying repositories for available binary packages ... Done!
#> * Querying repositories for available source packages ... Done!
#> Retrieving 'https://cran.rstudio.com/bin/windows/contrib/3.6/glue_1.3.1.zip' ...
#> 	OK [file is up to date]
#> Installing glue [1.3.1] from CRAN ...
#> 	OK (installed binary)
```

This step took ~10s to finish after the "OK" print. Then:

``` r
renv::snapshot()
#> * The lockfile is already up to date.
renv::status()
#> * The project is already synchronized with the lockfile.
library(glue)
#> 
#> Restarting R session...
#> 
#> * Project 'S:/tmp/bar' loaded. [renv 0.7.0-112]
renv::clean()
#> * No stale lockfiles were found.
#> * No temporary directories were found in the project library.
#> * No non-system packages were discovered in the system library.
#> The following packages are installed in the project library,
#> but appear to be no longer used in your project.
#> 
#> 	glue
#> 
#> These packages will be removed.
#> 
#> Do you want to proceed? [y/N]: y
#> * Removing package(s) from project library ...
#> Removing package 'glue' ... Done!
#> * The project has been cleaned.
```

I thought maybe it might be a cache thing, so I tried setting `RENV_PATHS_CACHE` to point to a directory on the same drive that the project was on: this seemingly had no effect.

**Comment by kevinushey** [comment]
Is there anything unique about these secondary drives? Are they regular local disks, network filesystems, or anything else?

Also worth confirming: are you using RStudio, or is this RGui or a plain R session in the terminal? If you're using RStudio, does switching to RGui make a difference?

You could try running a R profile and sharing the generated output so I could try and see what's going on as well:

``` r
Rprof("renv.Rprof")
# run slow code
Rprof(NULL)
```

You could then share that `renv.Rprof` file some way (e.g. using a downloadable link from OneDrive).

**Comment by mikmart** [comment]
Drives are regular local disks, one of them and SSD and another a hard drive.

I was only using RStudio, but I tested both RGui and just a terminal R session: I'm seeing the slowdown for all of them, but terminal session seems systematically a few seconds faster than the RStudio one.

``` r
# RStudio session on S:/
> system.time(renv::install(c("praise", "glue")))
Installing praise [1.0.0] ...
	OK (linked cache)
Installing glue [1.3.1] ...
	OK (linked cache)
   user  system elapsed 
   1.22    0.50   12.03 

# terminal session on S:/
> system.time(renv::install(c("praise", "glue")))
Installing praise [1.0.0] ...
        OK (linked cache)
Installing glue [1.3.1] ...
        OK (linked cache)
   user  system elapsed
   0.12    0.33   10.77

# terminal session on C:/
> system.time(renv::install(c("praise", "glue")))
Installing praise [1.0.0] ...
        OK (linked cache)
Installing glue [1.3.1] ...
        OK (linked cache)
   user  system elapsed
   0.11    0.22    0.39
```

I ran this code in a terminal session on the project on S:/

``` r
Rprof("renv.Rprof")
renv::install(c("praise", "glue"))
Rprof(NULL)
```

and here's the result: https://www.dropbox.com/s/6yfgbsqhdpfx7zw/renv.Rprof?dl=0

**Comment by mikmart** [comment]
I realized after reading the Q&A that the snapshot behaviour is expected. The packages I was installing weren't yet used in any script, so weren't being discovered as dependencies and thus not included in the lock file by taking a snapshot.

**Comment by kevinushey** [comment]
Thanks for generating the profile! Here's what I see (just the most important bit at the top):

```
> summaryRprof("~/Desktop/renv.rprof")
$by.self
                        self.time self.pct total.time total.pct
"file.rename"                9.70    93.27       9.70     93.27
```

For whatever reason, `file.rename()` is taking a very long time. I have a hypothesis...

Here's roughly what happens when `renv` installs a package:

1. `renv` sets up a temporary library to be used for newly-installed packages. This library lives within the R temporary directory, which I'm guessing is different from your project directory?

2. Then, `renv` installs packages into that library;

3. On success, `renv` will move (or copy) the packages from that temporary library into the project library.

My suspicion then is that `file.rename()` falls back to a copy + delete when attempting to move across disk drives, which can be very slow (especially since one here is not an SSD).

I think we could fix this in `renv` by using a temporary library within the project directory, rather than the session temporary directory.

**Comment by cderv** [comment]
> My suspicion then is that file.rename() falls back to a copy + delete when attempting to move across disk drives, which can be very slow (especially since one here is not an SSD).

On windows, we have this in the documentation for `?file.rename`
> This is subject to the limitations of the OS's corresponding system call (see something like man 2 rename on a Unix-alike): in particular in the interpretation of ‘file’: most platforms will not rename files from one file system to another. (**On Windows, file.rename can rename files but not directories across volumes.**) On platforms which allow directories to be renamed, typically neither or both of from and to must a directory, and if to exists it must be an empty directory.

This seems to confirm your diagnostic, right ?

**Comment by kevinushey** [comment]
I believe so -- I'll try to validate this on my Windows machine at home tomorrow as well.

Nonetheless, I've made a change in `renv` to ensure that we use a temporary directory within the target library for staging, as opposed to the R temporary directory. I believe this should resolve the issue that @mikmart is seeing.

I think this also matches what R itself does when performing its own formed of staged install (as supported in R 3.6.x).

**Comment by kevinushey** [comment]
Still, things may be slow if packages need to be copied to and from the cache -- we'll see if the change makes a difference, anyhow...

**Comment by mikmart** [comment]
I tried this also on my laptop, which has a recovery partition on the same single disk mapped to a different letter drive: I was experiencing the same delay in installs with 0.7.0-112.

However, just now I upgraded to 0.7.0-119 and the delay is gone---with and without caching! 🎉 Haven't tried on my desktop that has actual separate disks (accidentally triggered a large update going 🙄 ) but seems promising.
